pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+// and has the proper settings.
+func (t *Task) ensureSourceMigrationRegistry() error {
+	t.SrcRegistryResources = &MigrationRegistryResources{}
+	client, err := t.getSourceClient()
+	if err != nil {
+		return err
+	}
+	return t.ensureMigrationRegistry(client, t.SrcRegistryResources)
+}
+
+// Ensure the migration registry on the destination cluster has been created
+// and has the proper settings.
+func (t *Task) ensureDestinationMigrationRegistry() error {
+	t.DestRegistryResources = &MigrationRegistryResources{}
+	client, err := t.getDestinationClient()
+	if err != nil {
+		return err
+	}
+	return t.ensureMigrationRegistry(client, t.DestRegistryResources)
+}
+
+// Ensure the migration registry on the specified cluster has been created
+// and has the proper settings.
+func (t *Task) ensureMigrationRegistry(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
+	err := t.ensureRegistryCredentialsSecret(client, registryResources)
+	if err != nil {
+		return err
+	}
+
+	err = t.ensureRegistryImageStream(client, registryResources)
+	if err != nil {
+		return err
+	}
+
+	err = t.ensureRegistryDeploymentConfig(client, registryResources)
+	if err != nil {
+		return err
+	}
+	err = t.ensureRegistryService(client, registryResources)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Ensure the credentials secret for the migration registry on the specified cluster has been created
+func (t *Task) ensureRegistryCredentialsSecret(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
+	newSecret, err := t.buildRegistryCredentialsSecret(client)
+	if err != nil {
+		return err
+	}
+	foundSecret, err := t.getRegistryCredentialsSecret(client)
+	if err != nil {
+		return err
+	}
+	if foundSecret == nil {
+		registryResources.CredSecret = newSecret
+		err = client.Create(context.TODO(), newSecret)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.CredSecret = foundSecret
+	if !t.equalsRegistryCredentialsSecret(newSecret, foundSecret) {
+		t.updateRegistryCredentialsSecret(client, foundSecret)
+		err = client.Update(context.TODO(), foundSecret)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Build a credentials Secret as desired for the source cluster.
+func (t *Task) buildRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	secret := &kapi.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       t.PlanResources.MigPlan.GetCorrelationLabels(),
+			GenerateName: t.PlanResources.MigPlan.GetName() + "-registry-",
+			Namespace:    VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryCredentialsSecret(client, secret)
+	return secret, err
+}
+
+// Update a Registry credentials secret as desired for the specified cluster.
+func (t *Task) updateRegistryCredentialsSecret(client k8sclient.Client, secret *kapi.Secret) error {
+	credSecret, err := t.PlanResources.MigStorage.Spec.BackupStorageConfig.GetCredsSecret(t.Client)
```

Comment:
Two different clients. passed-in client is what we use to set/create src and destination resources, but t.Client here is specifically referencing the *controller* client, since that's where the MigStorage secret will be found. Otherwise this fails for remote clusters.

Reply:
Right, but I don't think we are actually using the client passed in as a param here (unless I'm missing something).

Reply:
Oh, right -- good catch. I initially added the client param before I realized I needed to use t.Client, and then when I updated the other code I forgot that it wasn't in the 'update' methods that the client param was normally used (the other update methods don't have a client param at all). I wonder why Go didn't complain about this (non-used method param).